core/src/main/scala/io/finch/Endpoint.scala

```diff
@@ -129,6 +129,13 @@ trait Endpoint[A] { self =>
     override def toString = self.toString
   }
 
+  def flatMap[B](fn: A => Endpoint[B]): Endpoint[B] = new Endpoint[B] {
+    override def apply(input: Input): Option[(Input, () => Future[Output[B]])] = for {
+      (r1, ao) <- self(input)
+      (r2, bo) <- fn(Await.result(ao()).value)(r1)
```

Comment:
Just a few thoughts on this by playing with types

If we can write Eval Future Output as M then
Endpoint A = Input -> Option (Input, M A)
And now we are given
f :: A -> Endpoint B = A -> (Input -> (Input, M B))

What we implement is
flatMap :: Endpoint A -> f -> Endpoint B

If we can make this function
g :: (Input, M A) -> Option (Input, M B)
we can make flatMap because

flatMap =
self(Input).map(g) :: Option Option (Input, M B)
-> Option (Input, M B) // join

(Input, M A)
-> (Input, M (Input -> Option (Input, M B))) // map
-> M (Option (Input, M B)) // M (Input -> Option(Input, M B)) `ap` (pure Input)
-> // stuck. how can I eliminate M?

I hope someone smaller than me helps me